diff -ruN bash-5.3.orig/Makefile.in bash-5.3/Makefile.in
--- bash-5.3.orig/Makefile.in	2025-05-16 14:56:31.000000000 -0400
+++ bash-5.3/Makefile.in	2026-03-17 17:12:39.786679384 -0400
@@ -559,7 +559,7 @@
 BUILTIN_C_SRC  = $(DEFSRC)/mkbuiltins.c $(DEFSRC)/common.c \
 		 $(DEFSRC)/evalstring.c $(DEFSRC)/evalfile.c \
 		 $(DEFSRC)/bashgetopt.c $(GETOPT_SOURCE)
-BUILTIN_C_OBJ  = $(DEFDIR)/common.o $(DEFDIR)/evalstring.o \
+BUILTIN_C_OBJ  = $(DEFDIR)/common.o $(DEFDIR)/histtime.o $(DEFDIR)/evalstring.o \
 		 $(DEFDIR)/evalfile.o $(DEFDIR)/bashgetopt.o
 BUILTIN_OBJS = $(DEFDIR)/alias.o $(DEFDIR)/bind.o $(DEFDIR)/break.o \
 	       $(DEFDIR)/builtin.o $(DEFDIR)/cd.o $(DEFDIR)/colon.o \
@@ -1839,6 +1839,7 @@
 $(DEFDIR)/caller.o: ${topdir}/bashintl.h ${LIBINTL_H} $(BASHINCDIR)/gettext.h
 $(DEFDIR)/cd.o: ${topdir}/bashintl.h ${LIBINTL_H} $(BASHINCDIR)/gettext.h
 $(DEFDIR)/common.c: ${topdir}/bashintl.h ${LIBINTL_H} $(BASHINCDIR)/gettext.h
+$(DEFDIR)/histtime.c: ${topdir}/bashintl.h ${LIBINTL_H} $(BASHINCDIR)/gettext.h
 $(DEFDIR)/complete.o: ${topdir}/bashintl.h ${LIBINTL_H} $(BASHINCDIR)/gettext.h
 $(DEFDIR)/declare.o: ${topdir}/bashintl.h ${LIBINTL_H} $(BASHINCDIR)/gettext.h
 $(DEFDIR)/enable.o: ${topdir}/bashintl.h ${LIBINTL_H} $(BASHINCDIR)/gettext.h
diff -ruN bash-5.3.orig/bashansi.h bash-5.3/bashansi.h
--- bash-5.3.orig/bashansi.h	2024-03-25 12:17:49.000000000 -0400
+++ bash-5.3/bashansi.h	2026-03-17 17:12:39.791786664 -0400
@@ -36,7 +36,7 @@
 #endif /* !HAVE_STDLIB_H */
 
 /* If bool is not a compiler builtin, prefer stdbool.h if we have it */
-#if !defined (HAVE_C_BOOL)
+#if !defined (HAVE_C_BOOL) && !(defined (__STDC_VERSION__) && __STDC_VERSION__ >= 202311L)
 #  if defined (HAVE_STDBOOL_H)
 #    include <stdbool.h>
 #  else
diff -ruN bash-5.3.orig/bashline.c bash-5.3/bashline.c
--- bash-5.3.orig/bashline.c	2025-06-25 15:53:44.000000000 -0400
+++ bash-5.3/bashline.c	2026-03-17 17:12:39.792789224 -0400
@@ -934,8 +934,8 @@
    command being entered (if no explicit argument is given), otherwise on
    a command from the history file. */
 
-#define VI_EDIT_COMMAND		"fc -e \"${VISUAL:-${EDITOR:-vi}}\""
-#define EMACS_EDIT_COMMAND	"fc -e \"${VISUAL:-${EDITOR:-emacs}}\""
+#define VI_EDIT_COMMAND		"fc -e \"${VISUAL:-${EDITOR:-$(command -v editor || echo vi)}}\""
+#define EMACS_EDIT_COMMAND	"fc -e \"${VISUAL:-${EDITOR:-$(command -v editor || echo emacs)}}\""
 #define POSIX_VI_EDIT_COMMAND	"fc -e vi"
 
 static int
diff -ruN bash-5.3.orig/builtins/Makefile.in bash-5.3/builtins/Makefile.in
--- bash-5.3.orig/builtins/Makefile.in	2024-11-11 11:51:48.000000000 -0500
+++ bash-5.3/builtins/Makefile.in	2026-03-17 17:12:39.787183344 -0400
@@ -154,12 +154,12 @@
 	  $(srcdir)/reserved.def $(srcdir)/pushd.def $(srcdir)/shopt.def \
 	  $(srcdir)/printf.def $(srcdir)/complete.def $(srcdir)/mapfile.def
 
-STATIC_SOURCE = common.c evalstring.c evalfile.c getopt.c bashgetopt.c \
+STATIC_SOURCE = common.c histtime.c evalstring.c evalfile.c getopt.c bashgetopt.c \
 		getopt.h 
 
 OFILES = builtins.o \
 	alias.o bind.o break.o builtin.o caller.o cd.o colon.o command.o \
-	common.o declare.o echo.o enable.o eval.o evalfile.o \
+	common.o histtime.o declare.o echo.o enable.o eval.o evalfile.o \
 	evalstring.o exec.o exit.o fc.o fg_bg.o hash.o help.o history.o \
 	jobs.o kill.o let.o mapfile.o \
 	pushd.o read.o return.o set.o setattr.o shift.o source.o \
@@ -237,6 +237,7 @@
 mkbuiltins.o: mkbuiltins.c
 builtins.o: builtins.c
 common.o: common.c
+histtime.o: histtime.c
 bashgetopt.o: bashgetopt.c
 getopt.o: getopt.c
 evalstring.o: evalstring.c          
@@ -484,7 +485,7 @@
 exit.o: $(BASHINCDIR)/unlocked-io.h
 fc.o: $(topdir)/bashtypes.h $(BASHINCDIR)/posixstat.h
 fc.o: $(topdir)/builtins.h $(topdir)/command.h $(srcdir)/bashgetopt.h
-fc.o: $(topdir)/bashhist.h $(topdir)/parser.h
+fc.o: $(topdir)/bashhist.h $(topdir)/parser.h $(srcdir)/histtime.h
 fc.o: $(topdir)/command.h ../config.h $(BASHINCDIR)/memalloc.h $(topdir)/error.h
 fc.o: $(topdir)/general.h $(topdir)/xmalloc.h $(BASHINCDIR)/maxpath.h
 fc.o: $(topdir)/quit.h $(topdir)/dispose_cmd.h $(topdir)/make_cmd.h $(topdir)/sig.h
@@ -527,7 +528,7 @@
 history.o: $(topdir)/bashtypes.h
 history.o: $(topdir)/command.h ../config.h $(BASHINCDIR)/memalloc.h
 history.o: $(topdir)/error.h $(topdir)/general.h $(topdir)/xmalloc.h
-history.o: $(topdir)/quit.h $(topdir)/dispose_cmd.h $(topdir)/make_cmd.h
+history.o: $(topdir)/quit.h $(topdir)/dispose_cmd.h $(topdir)/make_cmd.h $(srcdir)/histtime.h
 history.o: $(topdir)/subst.h $(topdir)/externs.h $(topdir)/sig.h  $(topdir)/parser.h
 history.o: ${BASHINCDIR}/filecntl.h $(topdir)/shell.h $(topdir)/syntax.h $(topdir)/unwind_prot.h
 history.o: $(topdir)/variables.h $(topdir)/conftypes.h $(topdir)/bashhist.h $(BASHINCDIR)/maxpath.h
@@ -730,6 +731,7 @@
 caller.o: ${topdir}/bashintl.h ${LIBINTL_H} $(BASHINCDIR)/gettext.h
 cd.o: ${topdir}/bashintl.h ${LIBINTL_H} $(BASHINCDIR)/gettext.h
 common.c: ${topdir}/bashintl.h ${LIBINTL_H} $(BASHINCDIR)/gettext.h
+histtime.c: ${topdir}/bashintl.h ${LIBINTL_H} $(BASHINCDIR)/gettext.h
 complete.o: ${topdir}/bashintl.h ${LIBINTL_H} $(BASHINCDIR)/gettext.h
 declare.o: ${topdir}/bashintl.h ${LIBINTL_H} $(BASHINCDIR)/gettext.h
 enable.o: ${topdir}/bashintl.h ${LIBINTL_H} $(BASHINCDIR)/gettext.h
diff -ruN bash-5.3.orig/builtins/cd.def bash-5.3/builtins/cd.def
--- bash-5.3.orig/builtins/cd.def	2024-09-28 16:25:22.000000000 -0400
+++ bash-5.3/builtins/cd.def	2026-03-17 17:12:39.790701744 -0400
@@ -322,6 +322,8 @@
 	  return (EXECUTION_FAILURE);
 	}
       lflag = 0;
+    } else if (strcmp(list->word->word, "...") == 0){
+      dirname = "../../";
     }
 #if defined (CD_COMPLAINS)
   else if (list->next)
diff -ruN bash-5.3.orig/builtins/fc.def bash-5.3/builtins/fc.def
--- bash-5.3.orig/builtins/fc.def	2024-05-06 11:55:30.000000000 -0400
+++ bash-5.3/builtins/fc.def	2026-03-17 17:12:39.793809584 -0400
@@ -23,7 +23,7 @@
 $BUILTIN fc
 $FUNCTION fc_builtin
 $DEPENDS_ON HISTORY
-$SHORT_DOC fc [-e ename] [-lnr] [first] [last] or fc -s [pat=rep] [command]
+$SHORT_DOC fc [-e ename] [-l0ntr] [first] [last] or fc -s [pat=rep] [command]
 Display or execute commands from the history list.
 
 fc is used to list or edit and re-execute commands from the history list.
@@ -35,7 +35,9 @@
   -e ENAME	select which editor to use.  Default is FCEDIT, then EDITOR,
 		then vi
   -l 	list lines instead of editing
+  -0	set command delimiter to '\0'
   -n	omit line numbers when listing
+  -t	print command timestamps
   -r	reverse the order of the lines (newest listed first)
 
 With the `fc -s [pat=rep ...] [command]' format, COMMAND is
@@ -83,6 +85,7 @@
 #include <readline/history.h>
 #include "bashgetopt.h"
 #include "common.h"
+#include "histtime.h"
 
 #if !defined (errno)
 extern int errno;
@@ -113,7 +116,7 @@
 /* fc builtin command (fix command) for Bash for those who
    like K*rn-style history better than csh-style.
 
-     fc [-e ename] [-nlr] [first] [last]
+     fc [-e ename] [-ntl0r] [first] [last]
 
    FIRST and LAST can be numbers specifying the range, or FIRST can be
    a string, which means the most recent command beginning with that
@@ -124,7 +127,9 @@
       mode, then vi.
 
    -l means list lines instead of editing.
+   -0 means set command delimiter to '\0'.
    -n means no line numbers listed.
+   -t means print command timestamps.
    -r means reverse the order of the lines (making it newest listed first).
 
      fc -e - [pat=rep ...] [command]
@@ -181,11 +186,11 @@
 }
 
 /* String to execute on a file that we want to edit. */
-#define FC_EDIT_COMMAND "${FCEDIT:-${EDITOR:-vi}}"
+#define FC_EDIT_COMMAND "${FCEDIT:-${EDITOR:-$(command -v editor || echo vi)}}"
 #if defined (STRICT_POSIX)
 #  define POSIX_FC_EDIT_COMMAND "${FCEDIT:-ed}"
 #else
-#  define POSIX_FC_EDIT_COMMAND "${FCEDIT:-${EDITOR:-ed}}"
+#  define POSIX_FC_EDIT_COMMAND "${FCEDIT:-${EDITOR:-$(command -v editor || echo ed)}}"
 #endif
 
 int
@@ -193,23 +198,23 @@
 {
   register int i;
   register char *sep;
-  int numbering, reverse, listing, execute;
+  int numbering, timestamp, reverse, listing, zero, execute;
   int histbeg, histend, last_hist, retval, opt, rh, real_last;
   FILE *stream;
   REPL *rlist, *rl;
-  char *ename, *command, *newcom, *fcedit;
+  char *ename, *command, *newcom, *fcedit, *histtimefmt, *timestr;
   HIST_ENTRY **hlist;
   char *fn;
 
   numbering = 1;
-  reverse = listing = execute = 0;
+  timestamp = reverse = listing = zero = execute = 0;
   ename = (char *)NULL;
 
   /* Parse out the options and set which of the two forms we're in. */
   reset_internal_getopt ();
   lcurrent = list;		/* XXX */
   while (fc_number (loptend = lcurrent) == 0 &&
-	 (opt = internal_getopt (list, ":e:lnrs")) != -1)
+	 (opt = internal_getopt (list, ":e:l0ntrs")) != -1)
     {
       switch (opt)
 	{
@@ -229,6 +234,14 @@
 	  execute = 1;
 	  break;
 
+	case '0':
+	  zero = 1;
+	  break;
+
+	case 't':
+	  timestamp = 1;
+	  break;
+
 	case 'e':
 	  ename = list_optarg;
 	  break;
@@ -465,6 +478,8 @@
 	}
     }
 
+  histtimefmt = get_string_value ("HISTTIMEFORMAT");
+
   for (i = reverse ? histend : histbeg; reverse ? i >= histbeg : i <= histend; reverse ? i-- : i++)
     {
       QUIT;
@@ -472,15 +487,27 @@
 	continue;
       if (numbering)
 	fprintf (stream, "%d", i + history_base);
-      if (listing)
+      if (listing && (!zero || numbering))
 	{
 	  if (posixly_correct)
 	    fputs ("\t", stream);
 	  else
 	    fprintf (stream, "\t%c", histdata (i) ? '*' : ' ');
 	}
+      if (timestamp)
+	{
+	  timestr = (histtimefmt && *histtimefmt) ? builtin_histtime (hlist[i], histtimefmt) : (char *)NULL;
+	  timestr = ((timestr && *timestr) ? timestr : "");
+	  fputs (timestr, stream);
+	}
       if (histline (i))
-	fprintf (stream, "%s\n", histline (i));
+	{
+	  fputs (histline (i), stream);
+	  if (zero)
+	    fputc ('\0', stream);
+	  else
+	    fputc ('\n', stream);
+	}
     }
 
   if (listing)
diff -ruN bash-5.3.orig/builtins/history.def bash-5.3/builtins/history.def
--- bash-5.3.orig/builtins/history.def	2025-04-07 17:56:59.000000000 -0400
+++ bash-5.3/builtins/history.def	2026-03-17 17:12:39.787805504 -0400
@@ -87,6 +87,7 @@
 #include <readline/history.h>
 #include "bashgetopt.h"
 #include "common.h"
+#include "histtime.h"
 
 #if !defined (errno)
 extern int errno;
@@ -106,10 +107,6 @@
 #define CFLAG	0x40
 #define DFLAG	0x80
 
-#ifndef TIMELEN_MAX
-#  define TIMELEN_MAX 128
-#endif
-
 int
 history_builtin (WORD_LIST *list)
 {
@@ -347,25 +344,6 @@
 #define histline(i) (hlist[(i)]->line)
 #define histdata(i) (hlist[(i)]->data)
 
-static char *
-histtime (HIST_ENTRY *hlist, const char *histtimefmt)
-{
-  static char timestr[TIMELEN_MAX];
-  time_t t;
-  struct tm *tm;
-
-  t = history_get_time (hlist);
-  tm = t ? localtime (&t) : 0;
-  if (t && tm)
-    strftime (timestr, sizeof (timestr), histtimefmt, tm);
-  else if (hlist->timestamp && hlist->timestamp[0])
-    snprintf (timestr, sizeof (timestr), _("%s: invalid timestamp"),
-	(hlist->timestamp[0] == '#') ? hlist->timestamp + 1: hlist->timestamp);
-  else
-    strcpy (timestr, "??");
-  return timestr;
-}
-
 static int
 display_history (WORD_LIST *list)
 {
@@ -403,7 +381,7 @@
 	{
 	  QUIT;
 
-	  timestr = (histtimefmt && *histtimefmt) ? histtime (hlist[i], histtimefmt) : (char *)NULL;
+	  timestr = (histtimefmt && *histtimefmt) ? builtin_histtime (hlist[i], histtimefmt) : (char *)NULL;
 	  printf ("%5d%c %s%s\n", i + history_base,
 		  histdata(i) ? '*' : ' ',
 		  ((timestr && *timestr) ? timestr : ""),
diff -ruN bash-5.3.orig/builtins/histtime.c bash-5.3/builtins/histtime.c
--- bash-5.3.orig/builtins/histtime.c	1969-12-31 19:00:00.000000000 -0500
+++ bash-5.3/builtins/histtime.c	2026-03-17 17:12:39.787991224 -0400
@@ -0,0 +1,50 @@
+/* histtime.c - a utility for history builtins */
+
+/* Copyright (C) 2025 Free Software Foundation, Inc.
+
+   This file is part of GNU Bash, the Bourne Again SHell.
+
+   Bash is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   Bash is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with Bash.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <config.h>
+
+#include <stdio.h>
+#include <string.h>
+
+#include "../bashintl.h"
+#include <readline/history.h>
+
+#ifndef TIMELEN_MAX
+#  define TIMELEN_MAX 128
+#endif
+
+char *
+builtin_histtime (HIST_ENTRY *hlist, const char *histtimefmt)
+{
+  static char timestr[TIMELEN_MAX];
+  time_t t;
+  struct tm *tm;
+
+  t = history_get_time (hlist);
+  tm = t ? localtime (&t) : 0;
+  if (t && tm)
+    strftime (timestr, sizeof (timestr), histtimefmt, tm);
+  else if (hlist->timestamp && hlist->timestamp[0])
+    snprintf (timestr, sizeof (timestr), _("%s: invalid timestamp"),
+        (hlist->timestamp[0] == '#') ? hlist->timestamp + 1: hlist->timestamp);
+  else
+    strcpy (timestr, "??");
+  return timestr;
+}
diff -ruN bash-5.3.orig/builtins/histtime.h bash-5.3/builtins/histtime.h
--- bash-5.3.orig/builtins/histtime.h	1969-12-31 19:00:00.000000000 -0500
+++ bash-5.3/builtins/histtime.h	2026-03-17 17:12:39.788136704 -0400
@@ -0,0 +1,21 @@
+/* histtime.h -- extern declarations for functions defined in histtime.c. */
+
+/* Copyright (C) 2025 Free Software Foundation, Inc.
+
+   This file is part of GNU Bash, the Bourne Again SHell.
+
+   Bash is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   Bash is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with Bash.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+extern char *builtin_histtime (HIST_ENTRY *, const char *);
diff -ruN bash-5.3.orig/config-bot.h bash-5.3/config-bot.h
--- bash-5.3.orig/config-bot.h	2023-03-29 09:24:54.000000000 -0400
+++ bash-5.3/config-bot.h	2026-03-17 17:12:39.791073744 -0400
@@ -193,4 +193,4 @@
 /******************************************************************/
 
 /* If you don't want bash to provide a default mail file to check. */
-/* #undef DEFAULT_MAIL_DIRECTORY */
+#undef DEFAULT_MAIL_DIRECTORY
diff -ruN bash-5.3.orig/config-top.h bash-5.3/config-top.h
--- bash-5.3.orig/config-top.h	2024-10-24 14:14:21.000000000 -0400
+++ bash-5.3/config-top.h	2026-03-17 17:12:39.791461624 -0400
@@ -58,7 +58,7 @@
 /* The default value of the PATH variable. */
 #ifndef DEFAULT_PATH_VALUE
 #define DEFAULT_PATH_VALUE \
-  "/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/sbin:."
+  "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
 #endif
 
 /* If you want to unconditionally set a value for PATH in every restricted
@@ -92,20 +92,20 @@
 #define DEFAULT_BASHRC "~/.bashrc"
 
 /* System-wide .bashrc file for interactive shells. */
-/* #define SYS_BASHRC "/etc/bash.bashrc" */
+#define SYS_BASHRC "/etc/bash.bashrc"
 
 /* System-wide .bash_logout for login shells. */
-/* #define SYS_BASH_LOGOUT "/etc/bash.bash_logout" */
+#define SYS_BASH_LOGOUT "/etc/bash.bash_logout"
 
 /* Define this to make non-interactive shells begun with argv[0][0] == '-'
    run the startup files when not in posix mode. */
-/* #define NON_INTERACTIVE_LOGIN_SHELLS */
+#define NON_INTERACTIVE_LOGIN_SHELLS
 
 /* Define this if you want bash to try to check whether it's being run by
    sshd and source the .bashrc if so (like the rshd behavior).  This checks
    for the presence of SSH_CLIENT or SSH2_CLIENT in the initial environment,
    which can be fooled under certain not-uncommon circumstances. */
-/* #define SSH_SOURCE_BASHRC */
+#define SSH_SOURCE_BASHRC
 
 /* Define if you want the case-toggling operators (~[~]) and the
    `capcase' variable attribute (declare -c). */
diff -ruN bash-5.3.orig/examples/startup-files/Bash_aliases bash-5.3/examples/startup-files/Bash_aliases
--- bash-5.3.orig/examples/startup-files/Bash_aliases	2019-12-10 14:56:36.000000000 -0500
+++ bash-5.3/examples/startup-files/Bash_aliases	2026-03-17 17:12:39.792085824 -0400
@@ -41,20 +41,20 @@
 { 
     local count="$1" i;
     shift;
-    for i in $(seq 1 "$count");
+    for i in $(_seq 1 "$count");
     do
         eval "$@";
     done
 }
 
 # Subfunction needed by `repeat'.
-seq ()
+_seq ()
 { 
     local lower upper output;
     lower=$1 upper=$2;
 
     if [ $lower -ge $upper ]; then return; fi
-    while [ $lower -le $upper ];
+    while [ $lower -lt $upper ];
     do
 	echo -n "$lower "
         lower=$(($lower + 1))
diff -ruN bash-5.3.orig/input.c bash-5.3/input.c
--- bash-5.3.orig/input.c	2025-02-09 16:41:29.000000000 -0500
+++ bash-5.3/input.c	2026-03-17 17:12:39.790261624 -0400
@@ -520,7 +520,7 @@
       if (nr == 0)
 	bp->b_flag |= B_EOF;
       else
-	bp->b_flag |= B_ERROR;
+	fatal_error("error reading input file: %s", strerror(errno));
       return (EOF);
     }
 
diff -ruN bash-5.3.orig/tests/builtins.right bash-5.3/tests/builtins.right
--- bash-5.3.orig/tests/builtins.right	2025-03-24 17:00:46.000000000 -0400
+++ bash-5.3/tests/builtins.right	2026-03-17 17:12:39.788414664 -0400
@@ -373,7 +373,7 @@
  exit [n]                                typeset [-aAfFgiIlnrtux] name[=value>
  export [-fn] [name[=value] ...] or ex>  ulimit [-SHabcdefiklmnpqrstuvxPRT] [>
  false                                   umask [-p] [-S] [mode]
- fc [-e ename] [-lnr] [first] [last] o>  unalias [-a] name [name ...]
+ fc [-e ename] [-l0ntr] [first] [last]>  unalias [-a] name [name ...]
  fg [job_spec]                           unset [-f] [-v] [-n] [name ...]
  for NAME [in WORDS ... ] ; do COMMAND>  until COMMANDS; do COMMANDS-2; done
  for (( exp1; exp2; exp3 )); do COMMAN>  variables - Names and meanings of so>
@@ -466,7 +466,7 @@
  exit [n]                                typeset [-aAfFgiIlnrtux] name[=value>
  export [-fn] [name[=value] ...] or ex>  ulimit [-SHabcdefiklmnpqrstuvxPRT] [>
  false                                   umask [-p] [-S] [mode]
- fc [-e ename] [-lnr] [first] [last] o>  unalias [-a] name [name ...]
+ fc [-e ename] [-l0ntr] [first] [last]>  unalias [-a] name [name ...]
  fg [job_spec]                           unset [-f] [-v] [-n] [name ...]
  for NAME [in WORDS ... ] ; do COMMAND>  until COMMANDS; do COMMANDS-2; done
  for (( exp1; exp2; exp3 )); do COMMAN>  variables - Names and meanings of so>
diff -ruN bash-5.3.orig/tests/history.right bash-5.3/tests/history.right
--- bash-5.3.orig/tests/history.right	2025-04-09 16:01:07.000000000 -0400
+++ bash-5.3/tests/history.right	2026-03-17 17:12:39.788632664 -0400
@@ -2,7 +2,7 @@
 history: usage: history [-c] [-d offset] [n] or history -anrw [filename] or history -ps arg [arg...]
 ./history.tests: line 19: history: cannot use more than one of -anrw
 ./history.tests: line 22: fc: -v: invalid option
-fc: usage: fc [-e ename] [-lnr] [first] [last] or fc -s [pat=rep] [command]
+fc: usage: fc [-e ename] [-l0ntr] [first] [last] or fc -s [pat=rep] [command]
     1  for i in one two three; do echo $i; done
     2  /bin/sh -c 'echo this is $0'
     3  ls
@@ -394,3 +394,42 @@
 
 
     5  echo three
+testting numbered + null-terminated...
+3	 echo -e 'x3
+x4
+x5
+' >/dev/null @null-terminator@
+2	 echo x2 >/dev/null @null-terminator@
+1	 echo x1 >/dev/null @null-terminator@
+
+testing null-terminated...
+echo -e 'x3
+x4
+x5
+' >/dev/null @null-terminator@
+echo x2 >/dev/null @null-terminator@
+echo x1 >/dev/null @null-terminator@
+
+testing numbered + timestamps + null-terminated...
+3	 [prefix +0000]	echo -e 'x3
+x4
+x5
+' >/dev/null @null-terminator@
+2	 [prefix +0000]	echo x2 >/dev/null @null-terminator@
+1	 [prefix +0000]	echo x1 >/dev/null @null-terminator@
+
+testing numbered + timestamps...
+3	 [prefix +0000]	echo -e 'x3
+x4
+x5
+' >/dev/null
+2	 [prefix +0000]	echo x2 >/dev/null
+1	 [prefix +0000]	echo x1 >/dev/null
+
+testing timestamps...
+	 [prefix +0000]	echo -e 'x3
+x4
+x5
+' >/dev/null
+	 [prefix +0000]	echo x2 >/dev/null
+	 [prefix +0000]	echo x1 >/dev/null
diff -ruN bash-5.3.orig/tests/history.tests bash-5.3/tests/history.tests
--- bash-5.3.orig/tests/history.tests	2025-04-09 16:01:00.000000000 -0400
+++ bash-5.3/tests/history.tests	2026-03-17 17:12:39.788809664 -0400
@@ -135,3 +135,4 @@
 ${THIS_SH} ./history7.sub
 ${THIS_SH} ./history8.sub
 ${THIS_SH} ./history9.sub
+${THIS_SH} ./history10.sub
diff -ruN bash-5.3.orig/tests/history10.sub bash-5.3/tests/history10.sub
--- bash-5.3.orig/tests/history10.sub	1969-12-31 19:00:00.000000000 -0500
+++ bash-5.3/tests/history10.sub	2026-03-17 17:12:39.788951944 -0400
@@ -0,0 +1,57 @@
+#   This program is free software: you can redistribute it and/or modify
+#   it under the terms of the GNU General Public License as published by
+#   the Free Software Foundation, either version 3 of the License, or
+#   (at your option) any later version.
+#
+#   This program is distributed in the hope that it will be useful,
+#   but WITHOUT ANY WARRANTY; without even the implied warranty of
+#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#   GNU General Public License for more details.
+#
+#   You should have received a copy of the GNU General Public License
+#   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+trap 'rm -f $HISTFILE' 0 1 2 3 6 15
+
+unset HISTIGNORE
+
+HISTFILE=$TMPDIR/foohist-$$
+
+# won't change
+export TZ=UTC
+HISTTIMEFORMAT=$'[prefix %z]\t'
+
+HISTCONTROL=ignorespace
+
+format-output() {
+  sed 's/\o0/ @null-terminator@\n/g'
+}
+
+set -o history
+
+echo x1 >/dev/null
+echo x2 >/dev/null
+
+echo -e 'x3
+x4
+x5
+' >/dev/null
+
+ echo testting numbered + null-terminated...
+ builtin fc -l0r | format-output
+
+ echo
+ echo testing null-terminated...
+ builtin fc -l0nr | format-output
+
+ echo
+ echo testing numbered + timestamps + null-terminated...
+ builtin fc -l0tr | format-output
+
+ echo
+ echo testing numbered + timestamps...
+ builtin fc -ltr | format-output
+
+ echo
+ echo testing timestamps...
+ builtin fc -lntr | format-output
