Index: lftp-4.9.2/m4/needtrio.m4
===================================================================
--- lftp-4.9.2.orig/m4/needtrio.m4
+++ lftp-4.9.2/m4/needtrio.m4
@@ -9,6 +9,7 @@ AC_DEFUN([LFTP_NEED_TRIO],[
       else
 
       AC_RUN_IFELSE([AC_LANG_SOURCE([[
+#include <stdio.h>
 	 int main()
 	 {
 	    unsigned long long x=0,x1;

From 573d06c04674413ba9013673d083d6a110036bd0 Mon Sep 17 00:00:00 2001
From: Jonas Jelten <jj@sft.lol>
Date: Thu, 14 Aug 2025 19:00:59 +0200
Subject: fix stack corruption in `cd` for relative paths due to added prefix length
Origin: https://github.com/lavv17/lftp/commit/573d06c04674413ba9013673d083d6a110036bd0.patch
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/lftp/+bug/2091440
Forwarded: https://github.com/lavv17/lftp/issues/734
Last-Update: 2025-08-29
---

when the current path is not absolute, the implementation prefixes
the target directory to change to with ~/
this is 2 more characters than allocated, hence we write beyond
the buffer boundary.

in order to fix this, we allocate two more characters for the reconstruction
buffer, so it can hold a relative path prefixed with ~/ as well.
---
 src/ftpclass.cc | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/ftpclass.cc b/src/ftpclass.cc
index 1ea9579f6..eb5d0186b 100644
--- a/src/ftpclass.cc
+++ b/src/ftpclass.cc
@@ -651,7 +651,7 @@ int Ftp::SendCWD(const char *path,const char *path_url,Expect::expect_t c)
       }
    } else {
       char *path1=alloca_strdup(path); // to split it
-      char *path2=alloca_strdup(path); // to re-assemble
+      char *path2=alloca_strdup2(path, 2); // to re-assemble + add space for potential ~/ prefix
       if(AbsolutePath(path)) {
 	 if(real_cwd && !strncmp(real_cwd,path,real_cwd.length())
 	 && path[real_cwd.length()]=='/') {
diff --git a/lib/libc-config.h.old b/lib/libc-config.h
index 33bdb73..000850c 100644
--- a/lib/libc-config.h.old
+++ b/lib/libc-config.h
@@ -179,7 +179,7 @@
 # endif
 
 /* Include our copy of glibc <sys/cdefs.h>.  */
-# include <cdefs.h>
+#include <sys/cdefs.h>
 
 /* <cdefs.h> __inline is too pessimistic for non-GCC.  */
 # undef __inline
